### PR TITLE
CLI: Update watch command

### PR DESCRIPTION
**Pull request by zshipko**
Adds parameters to `irmin watch` that allow the user to specify a command that should be executed or written to for each update.

### CODE REVIEW

File: src/irmin-unix/cli.ml [hunk 1]
```diff
@@ -535,14 +546,60 @@ let watch =
...
+                         (S.path * (S.contents * S.metadata) Irmin.Diff.t) list)
+                      =
+                    S.Tree.diff x y
+                  in
+                  match command with
+                  | h :: t ->
+                      let s =
+                        Fmt.str "%a"
+                          Irmin.Type.(
+                            pp_json
```
Comment (samoht) [inline src/irmin-unix/cli.ml:559]: Maybe use `[%typ: ...]` to simplify this expression?

File: src/irmin-unix/cli.ml [hunk 2]
```diff
@@ -535,14 +546,60 @@ let watch =
...
+                      in
+                      let make_proc () =
+                        let p =
+                          Lwt_process.open_process_out
+                            (h, Array.of_list (h :: t))
+                        in
+                        proc := Some p;
+                        p
+                      in
+                      let proc =
```
Comment (samoht) [inline src/irmin-unix/cli.ml:574]: can you extract this code in a subfunction? It's a bit hard to review what this is doing exactly here.

**Comment by samoht** [comment]
This is very cool! But if I understand correctly the watch command is a long-running command, right? 

Do you think we could have a mode where the watch command just persist the command somewhere on the disk (along with the last commit hash when they were called) and the writers will take care of calling the command when needed instead? It's probably a too large change for 3.0 but that'd be great idea for 4.0 :p

File: src/irmin-unix/cli.ml (hunk 2 above)
Comment (zshipko) [reply src/irmin-unix/cli.ml]: Sure, I also added some comments to make it easier to follow

**Comment by zshipko** [comment]
@samoht right now the watch command can be long-running or a one-off that is called for each update. I think that writing to disk and having another process check it should be possible, but maybe we can merge this version then I'll make another PR for those new changes?

**Comment by zshipko** [comment]
Rebased for https://github.com/mirage/irmin/pull/1603

**Pull request merged**